### PR TITLE
Fix #2292: Update BaseOperation#createOrReplace()

### LINKED ISSUE #2292: Update BaseOperation#createOrReplace
## Description

[BaseOperation#createOrReplace](https://github.com/fabric8io/kubernetes-client/blob/2d1a9d47d42e6aa37ba3cb0e70ede9429c93e392/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java#L397) should be updated to behave consistently with what's agreed on #2289 comments. (All exceptions except 409 conflict should be propagated for the POST/create operation)

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

Fix #2292 

+ Updated BaseOperation#createOrReplace to fallback to replace() after doing create() rather than get() -> replace()
+ Updated ResourceCompare#equals() to only compare entities present in provided resource when compared with
  resource from cluster.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [X] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [X] I have implemented unit tests to cover my changes
 - [X] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [X] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java [hunk 1]
```diff
@@ -395,24 +397,41 @@ public D createOrReplaceWithNew() throws KubernetesClientException {
...
-    if (item == null) {
+    if (itemToCreateOrReplace == null) {
       throw new IllegalArgumentException("Nothing to create.");
     }
 
-    if (Utils.isNullOrEmpty(name) && item instanceof HasMetadata) {
-      return withName(((HasMetadata)item).getMetadata().getName()).createOrReplace(item);
+    if (Utils.isNullOrEmpty(name)) {
+
+      return withName(itemToCreateOrReplace.getMetadata().getName()).createOrReplace(itemToCreateOrReplace);
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:413]: I'm guessing this recurses into this function

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java [hunk 2]
```diff
@@ -395,24 +397,41 @@ public D createOrReplaceWithNew() throws KubernetesClientException {
...
+        throw exception;
+      }
+
+      // Conflict; Do Replace
+      T itemFromServer = fromServer().get();
+      if (ResourceCompare.equals(itemFromServer, itemToCreateOrReplace)) {
+        // Nothing changed, ignore
+        return itemToCreateOrReplace;
+      } else {
+        KubernetesResourceUtil.setResourceVersion(itemToCreateOrReplace, resourceVersion);
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:432]: Are we sure we want to reuse the resourceVersion retrieved in line 416? Or maybe use the one of the latest version retrieved from the server (427)?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java (hunk 1 above)
Comment (rohanKanojia) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:413]: Yes, I think previously also it was like this.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java [hunk 3]
```diff
@@ -395,24 +397,41 @@ public D createOrReplaceWithNew() throws KubernetesClientException {
...
+    } catch (KubernetesClientException exception) {
+      if (exception.getCode() != HttpURLConnection.HTTP_CONFLICT) {
+        throw exception;
+      }
+
+      // Conflict; Do Replace
+      T itemFromServer = fromServer().get();
+      if (ResourceCompare.equals(itemFromServer, itemToCreateOrReplace)) {
+        // Nothing changed, ignore
+        return itemToCreateOrReplace;
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:430]: Will a PUT in the server of the same resource update its resource version (propagate event to watchers, etc.)? if so, maybe this is something expected from users invoking this function. WDYT?

File: kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/CreateOrReplaceResourceTest.java [hunk 4]
```diff
@@ -70,10 +75,21 @@ public void testResourceCreate() throws Exception {
...
+  @DisplayName("Should throw Exception on failed create")
+  void testResourceCreateFailure() {
+    // Given
+    server.expect().post().withPath("/api/v1/namespaces/test/pods").andReturn(HttpURLConnection.HTTP_BAD_REQUEST, new PodBuilder()
+      .withNewMetadata().withResourceVersion("12345").endMetadata().build()).once();
+    KubernetesClient client = server.getClient();
 
-    server.expect().post().withPath("/api/v1/namespaces/test/pods").andReturn(201, new PodBuilder()
+    // When
+    assertThrows(KubernetesClientException.class, () -> client.resource(new PodBuilder().withNewMetadata().withName("pod123").endMetadata().build()).createOrReplace());
```
Comment (manusa) [inline kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/CreateOrReplaceResourceTest.java:86]: You are getting sonar issues here. In order to make sure exception is thrown in the right place:
```suggestion
    Pod pod = client.resource(new PodBuilder().withNewMetadata().withName("pod123").endMetadata().build());
    // When
    assertThrows(KubernetesClientException.class, () -> pod.createOrReplace());
```

Same for the rest of warnings.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java (hunk 3 above)
Comment (rohanKanojia) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:430]: I checked and doing a PUT with same resource doesn't change resource version or add any event to watcher

**Comment by rohanKanojia** [comment]
retest this please

**Comment by manusa** [comment]
[merge]

**Comment by manusa** [comment]
[merge]

**Pull request merged**